### PR TITLE
feat: MSYS2 build script enhancements

### PR DESCRIPTION
**Pull request by offer-shmuely**
- stop on error
- support different branch's (not only main)
- add all the radio option on top of the script
- support re-run of the build (handle existing dir & git pull)
- add logs to follow the progress steps (especially in case of failure)

Fixes #

Summary of changes:

### CODE REVIEW

**Comment by offer-shmuely** [comment]
setting options:
image


final info:
image

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 1]
```diff
@@ -15,81 +41,109 @@ do
 	fi
 done
 
+STEP=1
 echo "=== Step $((STEP++)): Creating a directory for EdgeTX ==="
-mkdir ~/edgetx && cd ~/edgetx
+mkdir -p ~/edgetx && cd ~/edgetx
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:46]: home should exist, thus -p not required here.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 2]
```diff
@@ -15,81 +41,109 @@ do
...
+  git clone --recursive -b ${BRANCH_NAME} https://github.com/EdgeTX/edgetx.git ${SOURCE_DIR}
+  check_command $? "git clone --recursive -b ${BRANCH_NAME} https://github.com/EdgeTX/edgetx.git ${SOURCE_DIR}"
+  cd ${SOURCE_DIR}
+else
+  echo "=== Step $((STEP++)): updating EdgeTX source tree (${BRANCH_NAME} branch) from GitHub ==="
+  cd ${SOURCE_DIR}
+  check_command $? "cd ${SOURCE_DIR}"
+  git checkout ${BRANCH_NAME}
+  check_command $? "git checkout ${BRANCH_NAME}"
+  git pull
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:65]: Might be good idea to update the submodules here explicitly as well, e.g. with:
`git submodule update --init --recursive`

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 3]
```diff
@@ -15,81 +41,109 @@ do
...
+  check_command $? "git pull"
+fi
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
 echo "=== Step $((STEP++)): Creating build output directory ==="
-cd edgetx_main && mkdir build-output && cd build-output
+mkdir -p build-output && cd build-output
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:74]: If re-running the script, then a clean build output-directory would be in my opinion best for users, who need to use this script. Thus, I suggest to first wipe the build-output directory and then (re-)create it to have a clean state.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 4]
```diff
@@ -15,81 +41,109 @@ do
...
+check_command $? "mkdir -p build-output && cd build-output"
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
-echo "=== Step $((STEP++)): Running CMake for TX16S as an example ==="
-cmake -G "MSYS Makefiles" -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ -DPCB=X10 -DPCBREV=TX16S -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../
+echo "=== Step $((STEP++)): Running CMake for ${RADIO_TYPE} as an example ==="
+cmake -G "MSYS Makefiles" -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ -DPCB=X10 -DPCBREV=${RADIO_TYPE} -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:82]: If you want to be able to give radio type via variable, then also PCB should be dynamically changeable. Please note that some radio targets only use PCB and do NOT use PCBREV.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 5]
```diff
@@ -1,12 +1,38 @@
...
+## Bash script to show how to get EdgeTX source from GitHub,
 ## how to build TX16S firmware, Companion, Simulator, radio simulator
 ## library and how to create an installation package.
 ## Let it run as normal user in MSYS2 MinGW 64-bit console (blue icon).
 
-PAUSEAFTEREACHLINE="false"
-STEP=1
+# -----------------------------------------------------------------------------
+export BRANCH_NAME="main"  # main|2.8|2.9|...
+export RADIO_TYPE="TX16S"  # TX16S|X10|X10EXPRESS|X12S|XLITE|XLITES|X9LITE|X9LITES|X9D|X9D+|X9D+2019|X9E|X7|X7ACCESS|T16|T18||T12|TX12|COMMANDO8
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:10]: This needs further thought and refinement, as radio target is selected for some targets via a single PCB parameter, whereas some radio targets need two parameters specified - PCB and PCBREV.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 6]
```diff
@@ -15,81 +41,109 @@ do
...
+cmake -G "MSYS Makefiles" -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ -DPCB=X10 -DPCBREV=${RADIO_TYPE} -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../
+check_command $? "cmake -G MSYS Makefiles -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ -DPCB=X10 -DPCBREV=${RADIO_TYPE} -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../"
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
 echo "=== Step $((STEP++)): Running Make configure ==="
 make configure
+check_command $? "make configure"
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:91]: This step is required only for EdgeTX fw 2.8 and later and should be skipped for branches earlier than 2.8.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 7]
```diff
@@ -15,81 +41,109 @@ do
...
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
 echo "=== Step $((STEP++)): Building firmware binary ==="
 make -C arm-none-eabi -j`nproc` firmware
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:98]: For pre EdgeTX v2.8 branches, should use:
```
make -j`nproc` firmware
```

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 8]
```diff
@@ -15,81 +41,109 @@ do
...
+check_command $? "make -C arm-none-eabi -j`nproc` firmware"
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
 echo "=== Step $((STEP++)): Renaming firmware binary ==="
-mv arm-none-eabi/firmware.bin arm-none-eabi/fw_edgetx_main_tx16s_lua-ppmus-mode2_release.bin
+mv arm-none-eabi/firmware.bin arm-none-eabi/fw_${SOURCE_DIR}_${RADIO_TYPE}_lua-ppmus-mode2_release.bin
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:106]: Also here, the resulting binary is directly in `build-output` directory for pre EdgeTX v2.8 branches and not in `arm-none-eabi` subfolder.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 9]
```diff
@@ -1,12 +1,38 @@
...
-## Bash script to show how to get EdgeTX source from GitHub (main branch),
+## Bash script to show how to get EdgeTX source from GitHub,
 ## how to build TX16S firmware, Companion, Simulator, radio simulator
 ## library and how to create an installation package.
 ## Let it run as normal user in MSYS2 MinGW 64-bit console (blue icon).
 
-PAUSEAFTEREACHLINE="false"
-STEP=1
+# -----------------------------------------------------------------------------
+export BRANCH_NAME="main"  # main|2.8|2.9|...
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:9]: You might want to explicitly opt to limit to 2.8 and newer branches/tags only, as some build commands are to be handled differently for older firmware versions.

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 1 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:46]: how it was created?
I did not had one

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 2 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:65]: good idea
done

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 3 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:74]: good point
I also added the name of the radio to that dir

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 4 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:82]: not familiar with the actual build
can you point me to the list of options?

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 6 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:91]: so actually the script did not supported pre 2.8 build
so let comment it

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 7 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:98]: commented as 2.8 and above

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 8 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:106]: commented as 2.8 and above

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 9 above)
Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:9]: commented as 2.8 and above

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 10]
```diff
@@ -1,12 +1,45 @@
 #! /usr/bin/env bash
 
-## Bash script to show how to get EdgeTX source from GitHub (main branch),
+## Bash script to show how to get EdgeTX source from GitHub,
 ## how to build TX16S firmware, Companion, Simulator, radio simulator
 ## library and how to create an installation package.
 ## Let it run as normal user in MSYS2 MinGW 64-bit console (blue icon).
+## this script is working on any branch from 2.8 (icluded)
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:7]: Try "This script works properly only for branches stemming from EdgeTX v2.8 or newer."

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 4 above)
Comment (rotorman) [reply tools/msys2_fetch_and_build_all_tx16s.sh:91]: See the build-<manufacturer>.py scripts for hints under https://github.com/EdgeTX/edgetx/tree/main/tools

File: tools/msys2_fetch_and_build_all_tx16s.sh (hunk 5 above)
Comment (pfeerick) [reply tools/msys2_fetch_and_build_all_tx16s.sh:11]: Also, have a look at line 92 of this very script...
`-DPCB=X10 -DPCBREV=${RADIO_TYPE}`
... X10 is not valid for any of the B&W radios listed - only colorlcd radios, and not even all of them at that (i.e. not the NV14/EL18/PL18). 

Basically, you'll want to have those two parameters, and indicate that one essentially selects the radio family (PCB), and the second specifies the variant (PCBREV). 

This should give you a good idea as to the combinations:
https://github.com/EdgeTX/edgetx/blob/main/tools/build-gh.sh#L95-L178

Comment (offer-shmuely) [reply tools/msys2_fetch_and_build_all_tx16s.sh:11]: Ok
Will change that

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 11]
```diff
@@ -1,12 +1,147 @@
...
 ## library and how to create an installation package.
 ## Let it run as normal user in MSYS2 MinGW 64-bit console (blue icon).
+##
+## Note: This script works properly only for branches stemming from EdgeTX v2.8 or newer
+
+# -----------------------------------------------------------------------------
+export BRANCH_NAME="main"  # main|2.8|2.9|...
+export RADIO_TYPE="tx16s"  # x9lite|x9lites|x7|x7-access|t12|tx12|tx12mk2|boxer|t8|zorro|tlite|tpro|lr3pro|xlite|xlites|x9d|x9dp|x9dp2019|x9e|x9e-hall|x10|x10-access|x12s|t16|t18|tx16s|nv14|commando8|
+
+export BUILD_OPTIONS=""
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:14]: Add the custom build options here:
`export BUILD_OPTIONS="-DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES"`

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 12]
```diff
@@ -15,81 +150,111 @@ do
...
+check_command $? "mkdir -p ${BUILD_OUTPUT_DIR} && cd ${BUILD_OUTPUT_DIR}"
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
-echo "=== Step $((STEP++)): Running CMake for TX16S as an example ==="
-cmake -G "MSYS Makefiles" -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ -DPCB=X10 -DPCBREV=TX16S -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../
+echo "=== Step $((STEP++)): Running CMake for ${RADIO_TYPE} as an example ==="
+cmake -G "MSYS Makefiles" -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ ${BUILD_OPTIONS} -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DHELI=NO -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:193]: Remove the custom build options here and move them to line 14:
`cmake -G "MSYS Makefiles" -Wno-dev -DCMAKE_PREFIX_PATH=$HOME/5.12.9/mingw73_64 -DSDL_LIBRARY_PATH=/mingw64/bin/ ${BUILD_OPTIONS} -DCMAKE_BUILD_TYPE=Release ../`

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 13]
```diff
@@ -15,81 +150,111 @@ do
...
 make -C arm-none-eabi -j`nproc` firmware
+check_command $? "make -C arm-none-eabi -j`nproc` firmware"
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
   read
 fi
 
 echo "=== Step $((STEP++)): Renaming firmware binary ==="
-mv arm-none-eabi/firmware.bin arm-none-eabi/fw_edgetx_main_tx16s_lua-ppmus-mode2_release.bin
+mv arm-none-eabi/firmware.bin arm-none-eabi/fw_${SOURCE_DIR}_${RADIO_TYPE}_lua-ppmus-mode2_release.bin
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:217]: Here is a bug, please run the script to test and see this: folder `arm-none-eabi/fw_${SOURCE_DIR}` does not exist (SOURCE_DIR is evaluated to `${HOME}/edgetx/edgetx_${BRANCH_NAME}`). Try instead:
`mv arm-none-eabi/firmware.bin arm-none-eabi/fw_${RADIO_TYPE}_lua-ppmus-mode2_release.bin`

The same applies to the next line as well.

File: tools/msys2_fetch_and_build_all_tx16s.sh [hunk 14]
```diff
@@ -15,81 +150,111 @@ do
...
+check_command $? "make -C native installer"
 if [[ $PAUSEAFTEREACHLINE == "true" ]]; then
   echo "Step finished. Please check the output above and press Enter to continue or Ctrl+C to stop."
 fi
 
-echo "Finished. You can find the TX16S firmware in ~/edgetx/edgetx_main/build-output/arm-none-eabi"
-echo "and Companion installer packet in ~/edgetx/edgetx_main/build-output/native/companion"
+echo "Finished."
+echo ""
+echo "firmware (${RADIO_TYPE}): ${SOURCE_DIR}/${BUILD_OUTPUT_DIR}/arm-none-eabi/fw_${SOURCE_DIR}_${RADIO_TYPE}_lua-ppmus-mode2_release.bin"
```
Comment (rotorman) [inline tools/msys2_fetch_and_build_all_tx16s.sh:257]: Here similarly to lines 217 & 218 above - `arm-none-eabi/fw_${SOURCE_DIR}_$` evaluates to subfolder that does not exist.

**Review by rotorman — CHANGES_REQUESTED** [review]
Please see some further comments, especially regarding `fw_${SOURCE_DIR}` non-existent subfolder file move bug.

**Comment by offer-shmuely** [comment]
Thanks for your very good comments!

**Pull request merged**